### PR TITLE
ZAI function-call seam (PHP 8)

### PR DESCRIPTION
**Pull request by SammyK**
### Description

This PR adds a Zend Abstract Interface (ZAI) seam for safely calling (sandboxed) PHP functions from C in PHP 8:

- `bool zai_call_function_literal(const char *name, zval *retval, ...);`
- `bool zai_call_function(const char *name, size_t name_len, zval *retval, ...);`
- `bool zai_call_function_ex(const char *name, size_t name_len, zval *retval, int argc, ...);`

This seam will initially be used to call the [userland deferred-loading function](https://github.com/DataDog/dd-trace-php/blob/5b8ea36/ext/php8/php8/engine_hooks.c#L127-L160) in PHP 8.

This PR also includes a new macro for asserting lowercase strings (`zai_assert/zai_assert.h`):

```c
zai_assert_is_lower(str, "String must be lowercase.");
```

As with `assert()`, `zai_assert_is_lower()` will only abort on debug builds of PHP.

### Readiness checklist
- [x] (only for Members) Changelog has been added to the release document.
- [x] Tests added for this feature/bug.

### Reviewer checklist
- [ ] Appropriate labels assigned.
- [ ] Milestone is set.
- [ ] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: zend_abstract_interface/functions/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+                                    VERSION ${PROJECT_VERSION})
+
+add_library(Zai::Functions ALIAS zai_functions)
+
+if (${BUILD_ZAI_TESTING})
+  add_subdirectory(tests)
+endif()
+
+# This copies the include files when `install` is ran
+# TODO: How to make this zai/functions.h?
```
Comment (morrisonlevi) [inline zend_abstract_interface/functions/CMakeLists.txt:22]: To answer your question on making it `zai/functions.h`, you would set the `DESTINATION` to `"${CMAKE_INSTALL_INCLUDEDIR}/zai"` but to make that _also_ work prior to install (the common case for us) you need to adjust the directory structure so that it lives in `zai/functions.h`.

File: zend_abstract_interface/functions/php8/functions.c [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+#ifndef NDEBUG
+static bool z_is_lower(const char *str) {
+    char *p = (char *)str;
+    while (*p) {
+        if (isalpha(*p) && !islower(*p)) return false;
+        p++;
+    }
+    return true;
+}
+#endif
```
Comment (morrisonlevi) [inline zend_abstract_interface/functions/php8/functions.c:15]: I think this is missing an include for `<ctype.h>` when it's a debug build.

File: zend_abstract_interface/functions/tests/php8/functions.cc [hunk 3]
```diff
@@ -0,0 +1,381 @@
+extern "C" {
+#include "functions/functions.h"
+#include "zai_sapi/zai_sapi.h"
```
Comment (morrisonlevi) [inline zend_abstract_interface/functions/tests/php8/functions.cc:3]: These should use `#include <...>` right? I think that's what we decided?

File: zend_abstract_interface/sandbox/php8/sandbox.c [hunk 4]
```diff
@@ -4,7 +4,30 @@ extern inline void zai_sandbox_open(zai_sandbox *sandbox);
...
 extern inline void zai_sandbox_error_state_backup(zai_error_state *es);
-extern inline void zai_sandbox_error_state_restore(zai_error_state *es);
+
+/* In PHP 8, we cannot extern inline zai_sandbox_error_state_restore() because
+ * zend_string_release() is static which generates the following compiler
+ * warning:
+ *
+ *   "warning: 'zend_string_release' is static but used in inline function
+ *    'zai_sandbox_error_state_restore' which is not static"
+ */
```
Comment (morrisonlevi) [inline zend_abstract_interface/sandbox/php8/sandbox.c:14]: Just to document some history: I tried talking Nikita into allowing me to fix these in php-src but he didn't see the value in doing `inline` + `extern inline` vs allowing the compiler to do it via `static inline`. Plus there are `zend_always_inline` cases which have _theoretically_ been put there because of performance reasons, and we wouldn't want them to accidentally _not_ get inlined (although "always" is technically a lie anyway). I didn't feel like doing the necessary performance A/B work to demonstrate there weren't regressions and dropped it.

tl;dr It'll probably never get fixed upstream.

File: zend_abstract_interface/functions/tests/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+#[[ TODO When we have a consistent API across all PHP versions with ZAI
+    data-structure shims, move these tests out of PHP-versioned directories.
+ ]]
+add_executable(functions "${PHP_VERSION_DIRECTORY}/functions.cc")
+
+target_link_libraries(functions PUBLIC catch2_main Zai::Sapi Zai::Functions)
+
+file(COPY ${CMAKE_CURRENT_SOURCE_DIR}/stubs
+     DESTINATION ${CMAKE_CURRENT_BINARY_DIR})
```
Comment (morrisonlevi) [inline zend_abstract_interface/functions/tests/CMakeLists.txt:9]: I understand why the copy is necessary, but I don't like it for some reason I can't place. Did you have a similar feeling or is it just me?

**Review by morrisonlevi — COMMENTED** [review]
I still have more review to do, but posting these comments for now.

File: zend_abstract_interface/functions/CMakeLists.txt (hunk 1 above)
Comment (SammyK) [reply zend_abstract_interface/functions/CMakeLists.txt:22]: Ah yes, I remember our conversation on this a while back. I believe we decided to not put all the headers in the same directory for some reason, but I can't remember the reason. I think we decided to keep it `thing/thing.h` until we could figure out a proper solution that we liked. Since this is also something that impacts `libddprof`, I would love to sync up with you on this again to iron out a direction we want to go.

File: zend_abstract_interface/functions/tests/php8/functions.cc (hunk 3 above)
Comment (SammyK) [reply zend_abstract_interface/functions/tests/php8/functions.cc:3]: I think using `<>` makes since from the perspective of including these from the tracer since they live outside of the tracer project. But since these are tests for testing _within_ the ZAI project, `""` seems to make more sense to me.

In general, this feels most natural to me:
- `<>` = external project header
- `""` = internal project header

WDYT?

File: zend_abstract_interface/functions/tests/CMakeLists.txt (hunk 5 above)
Comment (SammyK) [reply zend_abstract_interface/functions/tests/CMakeLists.txt:9]: The only quirky thing that I've experienced with this copy is that if you make a change to a stub, you basically have to blow away the build dir and start over again (at least that's what I do with my primitive `cmake` skills. 😆) But I'm not sure how else we'd get the stubs copied over for the tests without doing a `COPY`. Of course I'm totally open to ideas if you've got a better solution!

File: zend_abstract_interface/functions/php8/functions.c [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    zai_sandbox sandbox;
+    zai_sandbox_open(&sandbox);
+
+    zend_try {
+        zend_result result = zend_call_function(&fci, &fcc);
+        /* An unhandled exception will not result in a zend_bailout if there is
+         * an active execution context. This is a failed call if an exception
+         * was thrown. The sandbox will clean up our mess when it closes.
+         */
+        ret = (result == SUCCESS && !EG(exception));
```
Comment (labbati) [inline zend_abstract_interface/functions/php8/functions.c:115]: Not a change required for this PR, but rather an observation for the design of future zai modules. Do you think that a binary return state is enough to communicate the outcome? Or it would make sense to have at least 3 states:

1. `SUCCESS`;
2. error because of something you did wrong while developing (e.g. calling with a `NULL` return value);
3. runtime error (e.g. and exception because the database is not reachable).

File: zend_abstract_interface/functions/tests/php8/functions.cc [hunk 7]
```diff
@@ -0,0 +1,381 @@
+extern "C" {
```
Comment (labbati) [inline zend_abstract_interface/functions/tests/php8/functions.cc:1]: @SammyK I really love this testing infrastructure. You made an awesome job!

**Review by labbati — COMMENTED** [review]
Great work @SammyK, I asked a few questions, but as always mostly for me to learn things :)

File: zend_abstract_interface/functions/php8/functions.c (hunk 2 above)
Comment (SammyK) [reply zend_abstract_interface/functions/php8/functions.c:15]: Nice catch. Since this is a pretty common use case in the Zend Engine, and it is [already being used for methods & class lookups](https://github.com/DataDog/dd-trace-php/blob/f4b5065/zend_abstract_interface/methods/php5/methods.c#L7-L16) as well, I was thinking of extracting this into `zai_assert/zai_assert.h` as something like this:

```c
zai_assert_is_lower(str, "String must be lowercase.");
```

Do you think it's worth adding that as part of this PR?

File: zend_abstract_interface/functions/tests/CMakeLists.txt (hunk 5 above)
Comment (morrisonlevi) [reply zend_abstract_interface/functions/tests/CMakeLists.txt:9]: I think you need to add the copied file as a dependency of the test using `add_dependency`, though I'm not sure a `COPY` is recognized as a "factory" (I use quotes because I'm not sure what the proper CMake terminology is).

File: zend_abstract_interface/functions/functions.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ * The long-term goal is to provide ZAI data-structure shims so that the
+ * following APIs are the same across all PHP versions. For now we will provide
+ * ZAI seams using native Zend Engine data structures.
+ */
+
+/* Calls a function with 'argc' number of zval arguments. Caller owns the
+ * arguments and the 'retval'. Caller is responsible for freeing any refcounted
+ * arguments. Caller must provide a pointer to a stack-allocated zval for the
+ * return value, 'retval'. In error cases, 'retval' will be set to IS_UNDEF.
+ * Caller must dtor the 'retval' after the call:
```
[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/functions.h:18]: I'm not sure I understand what you're suggesting in terms of the refcount handling, but we certainly can (and will 🙂) add an abstraction layer/wrapper at the tracer level. (For example, auto-flushing on PHP 5 wraps the method-call seam with an [API that ignores the retval](https://github.com/DataDog/dd-trace-php/blob/5275a4f/ext/php5/auto_flush.c#L9-L17).)

On the refcounting bit - I think the comment might be a bit confusing the way it is worded.

- Basically the caller 100% owns the args (ctoring/passing/dtoring).
- The caller also owns the space for the retval (ideally stack-allocated).
- The only memory `zai_call_function` allocates is the space to hold the params to pass to the function and it is freed before returning to the caller. (As a side note we might want to consider a hard-limit on the number of args so that we can stack-allocate the params. 🤔)
- The retval may or may not need to be dtored depending on whether or not the PHP function returns a refcounted value (e.g. array, object, string, etc). But that should be known to the caller so the caller can make that choice what to do with the retval.

File: zend_abstract_interface/functions/php8/functions.c [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    /* For consistency with zend_call_function(), this wrapper also initializes
+     * the retval to IS_UNDEF so that the caller can expect an undefined retval
+     * in the error cases. Past wrappers of zend_call_function() did not mimic
+     * this behavior and that lead to a SIGSEGV.
+     *
+     * https://github.com/php/php-src/blob/PHP-8.0.3/Zend/zend_execute_API.c#L673
+     */
+    ZVAL_UNDEF(retval);
+
+    if (!name || !name_len) return false;
```
[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/php8/functions.c:29]: For simple NULL checks like this, I find that going sans curly brackets is more readable, but if there's a strong technical argument for adding the curly brackets, I'm certainly open to changing it. I'm curious if @morrisonlevi has any thoughts on this as well?

File: zend_abstract_interface/functions/php8/functions.c [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+
+    /* Functions cannot be called outside of a request context.
+     * PG(modules_activated) indicates that all of the module RINITs have been
+     * called and we are in a request context.
+     */
+    if (!PG(modules_activated)) return false;
+
+    assert(z_is_lower(name) && "Function names must be lowercase.");
```
[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/php8/functions.c:37]: > On the other hand - unless PHP does some magic that I am not aware of - assert would cause the program execution to abort.

This is only true for debug builds of PHP because [`NDEBUG` is defined for non-debug builds of PHP](https://github.com/php/php-src/blob/c40231afbf85e0828a6c56d289cb823d65d03a54/main/php.h#L120-L126). To validate this, we [run the ZAI tests on non-debug builds](https://github.com/DataDog/dd-trace-php/blob/6f1f4b2/.circleci/config.yml#L1083) to ensure there is no abort happening for [these assertions](https://github.com/DataDog/dd-trace-php/blob/6f1f4b2/zend_abstract_interface/methods/tests/php5/methods.cc#L63-L87).

> Since the user is already "forced" to handle failure due to the return false above, do you think it would make sense here to just return false here are well?

This is the behavior on non-debug builds of PHP since the hash table lookup will fail if the function name is not lowercase (and then return `false`).

File: zend_abstract_interface/functions/php8/functions.c [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    ZSTR_ALLOCA_FREE(zsname, use_heap);
+
+    /* "function %s() does not exist" */
+    if (!func) return false;
+
+    zend_fcall_info_cache fcc = {0};
+    fcc.function_handler = func;
+
+    zend_fcall_info fci = {0};
+    fci.size = sizeof(zend_fcall_info);
```
[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/php8/functions.c:70]: I believe this is a relic from an [18 year old refactoring](https://github.com/php/php-src/commit/92b4013e8d6fb3a92f958e9c01abf824bbdb4aaa). Today it is only [accessed in debug builds](https://github.com/php/php-src/blob/dd86987b2cbead2af1e95460cc5cebe03e83f736/Zend/zend_execute_API.c#L709).

I think we could probably remove it from core, but I'm curious what @bwoebi thinks.

File: zend_abstract_interface/functions/php8/functions.c (hunk 2 above)
Comment (morrisonlevi) [reply zend_abstract_interface/functions/php8/functions.c:15]: Sounds good to me; let's do it.

File: zend_abstract_interface/functions/functions.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (morrisonlevi) [reply zend_abstract_interface/functions/functions.h:18]: > [W]e might want to consider a hard-limit on the number of args so that we can stack-allocate the params. 🤔

Allowing the caller to manage the memory so it can stack allocate with exactly the right number without any limits seems like it might be better. I think it may lose some expressiveness, though? Is that why you went this way?

File: zend_abstract_interface/functions/php8/functions.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (morrisonlevi) [reply zend_abstract_interface/functions/php8/functions.c:29]: I am fine with short versions of this pattern, as long as the whole thing fits on a single line:

    if (conditions) return;
    
I'm also fine with `break` and `continue`.

File: zend_abstract_interface/functions/php8/functions.c [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+
+        for (uint32_t i = 0; i < (uint32_t)argc; ++i) {
+            arg = va_arg(argv, zval *);
+            if (!arg) {
+                /* Only dtor the zvals that were successfully copied */
+                fci.param_count = i;
+                zend_fcall_info_args_clear(&fci, /* free_mem */ true);
+                return false;
+            }
+            ZVAL_COPY(&fci.params[i], arg);
```
[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/php8/functions.c:98]: > why should we copy the zval to invoke a function instead of refcounting +1?

[It does both](https://github.com/php/php-src/blob/dd86987b2cbead2af1e95460cc5cebe03e83f736/Zend/zend_types.h#L1259-L1262). We need to copy the `zval` because we've allocated heap memory for the `params` on `zend_fcall_info` and we need to copy the `zval`s over to it. But after the above discussion about avoiding the heap allocation altogether, I might refactor this a bit.

File: zend_abstract_interface/functions/php8/functions.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (bwoebi) [reply zend_abstract_interface/functions/php8/functions.c:70]: No, `ZEND_FCI_INITIALIZED()` also accesses it - technically core could change that to a `zend_bool`, but seems sort of pointless to do that arbitrary ABI break. If we really strive to remove it, we'd need to devise some alternative handling for nullable callables in zpp (see `'f'` handling of `zend_parse_arg_impl`). (Which makes sense, but it's not as trivial and extension authors need to adapt to the new handling.)

File: zend_abstract_interface/functions/functions.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (bwoebi) [reply zend_abstract_interface/functions/functions.h:18]: I'm not sure why we need a hard limit for stack allocated arrays? simple variable length array: `zval params[argc];`. (At least I assume that we can use C99 here?)

File: zend_abstract_interface/functions/php8/functions.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (bwoebi) [reply zend_abstract_interface/functions/php8/functions.c:29]: I'm fine with non-block ifs as long as they're strictly single line.

File: zend_abstract_interface/functions/php8/functions.c (hunk 6 above)
Comment (SammyK) [reply zend_abstract_interface/functions/php8/functions.c:115]: > Do you think that a binary return state is enough to communicate the outcome?

It is not.

> Or it would make sense to have at least 3 states:

Very good point. [ZAI error types are on the wishlist](https://github.com/DataDog/dd-trace-php/blob/f417c10/zend_abstract_interface/sandbox/sandbox.h#L7-L12). ([APMPHP-614](https://datadoghq.atlassian.net/browse/APMPHP-614)) They will likely be tightly coupled to the sandbox so that exception and error states can be moved out of the PHP global space and into a locally-managed error state.

The return states would be similar to what you proposed. Off the top of my head, this will include the following basic return states:

0. **Success**
1. **Bailout**; Allows the caller to clean up before bubbling up a `zend_bailout`.
2. **Failure**; This can have a number of generic sub types (e.g. exception thrown, error raised, NULL argument, etc) and API-specific types (e.g. method does not exist).

File: zend_abstract_interface/functions/tests/php8/functions.cc (hunk 7 above)
Comment (SammyK) [reply zend_abstract_interface/functions/tests/php8/functions.cc:1]: Thanks! The ZAI SAPI test runner piggybacks off of the original Catch2 work that @morrisonlevi [contributed back in October 2020](https://github.com/DataDog/dd-trace-php/commit/aafae27f64a7475ed77d502f3d06ad0510e40a48). #TeamWork 🙂

File: zend_abstract_interface/functions/php8/functions.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (bwoebi) [reply zend_abstract_interface/functions/php8/functions.c:98]: I'm not sure how that `because` is related (the only thing is, if you do not copy you cannot use zend_fcall_info_args_clear below. But that's not an issue...)? In fact zend_call_function itself already does a copy. We do not need a copy nor RC increase here at all.

File: zend_abstract_interface/functions/php8/functions.c (hunk 2 above)
Comment (SammyK) [reply zend_abstract_interface/functions/php8/functions.c]: Added in 4e194c58.

File: zend_abstract_interface/functions/functions.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/functions.h:18]: > I'm not sure why we need a hard limit for stack allocated arrays? simple variable length array: zval params[argc];. (At least I assume that we can use C99 here?)

Indeed we can; we use C99 and C11. Thanks for the tip. I've refactored the param handling in 8ae546bf.

[the remark this answers is not in the document]
Comment (morrisonlevi) [reply zend_abstract_interface/functions/functions.h:18]: For what it is worth, C11 moved VLA's to a conditional feature instead of mandatory. The compiler community seems to want to get rid of this feature. I think I still might prefer moving it to the caller, where the size is known.

File: zend_abstract_interface/functions/php8/functions.c (hunk 12 above)
[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/php8/functions.c]: > I'm not sure how that because is related (the only thing is, if you do not copy you cannot use zend_fcall_info_args_clear below. But that's not an issue...)?

Indeed, the heap allocation for the params was left over from expanding `zend_fcall_info_argv`, so once I removed that and the `zend_fcall_info_args_clear` calls in 8ae546bf, that simplified the code quite a bit.

**Fun fact:** I originally just expanded `zend_fcall_info_argv` in place just so that I could NULL-check the args, but after this discussion, I think we're onto something even better. :)

> In fact zend_call_function itself already does a copy. We do not need a copy nor RC increase here at all.

I didn't realize that the args were being copied and having the refcount increased from within `zend_call_function`. Thanks for the tip!

I'm not sure how we would take the args from the variable arguments `va_arg` and use them in the `zend_fcall_info` params without [some kind of copy](https://github.com/DataDog/dd-trace-php/blob/8ae546bf5e42a61a3e0e0dbe475a7b1e4644253c/zend_abstract_interface/functions/php8/functions.c#L75). But I may have misunderstood what you meant by not needing to copy?

[the remark this answers is not in the document]
Comment (bwoebi) [reply zend_abstract_interface/functions/php8/functions.c]: I meant the ZVAL_COPY (vs bare assignment) - so that's fine now :-)

File: zend_abstract_interface/functions/php8/functions.c [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+     * 'zend_fcall_info' will always be zeroed out in our case.
+     */
+    zval params[argc];
+    if (argc > 0) {
+        va_list argv;
+        va_start(argv, argc);
+        for (uint32_t i = 0; i < (uint32_t)argc; ++i) {
+            zval *arg = va_arg(argv, zval *);
+            if (!arg) return false;
+            params[i] = *arg;
```
Comment (bwoebi) [inline zend_abstract_interface/functions/php8/functions.c:75]: The canonical way to assign a zval would be actually `ZVAL_COPY_VALUE()`. This copies the zend_value as two individual 32 bit assignments on 32 bit environments (I do not know why. Perhaps that's completely irrelevant?!?) and notably does _not_ copy u2.

File: zend_abstract_interface/functions/functions.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/functions.h:18]: > I think I still might prefer moving it to the caller, where the size is known.

Do you see a significant advantage for moving this to the caller vs using a variable length array?

Although this might save us some assignments, my main concern is that the API will not be as expressive which would make it easier pass the params the wrong way. Since one of the main purposes of the ZAI API is to increase stability by reducing the chance of using the API incorrectly, my preference would be to keep it as is and reduce the cognitive load of using the API.

[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/functions.h:18]: Although, [UBSan does not like](https://app.circleci.com/pipelines/github/DataDog/dd-trace-php/5787/workflows/560dc271-42ab-4d60-8816-d46b0e29260e/jobs/435966/steps) our VLA with zero length (calling a function without args).

```
/home/circleci/datadog/zend_abstract_interface/functions/php8/functions.c:68:10: runtime error: variable length array bound evaluates to non-positive value 0
```

I'll need to fix this tomorrow. But 🎉  for having the `*San`s catch this in CI!

[the remark this answers is not in the document]
Comment (labbati) [reply zend_abstract_interface/functions/functions.h:18]: Thanks for the clarification! :)

File: zend_abstract_interface/functions/php8/functions.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (labbati) [reply zend_abstract_interface/functions/php8/functions.c:59]: No need of course :) thanks

File: zend_abstract_interface/functions/php8/functions.c (hunk 6 above)
Comment (labbati) [reply zend_abstract_interface/functions/php8/functions.c:94]: Thanks @SammyK for clarifying

File: zend_abstract_interface/functions/php8/functions.c (hunk 13 above)
Comment (SammyK) [reply zend_abstract_interface/functions/php8/functions.c]: TIL! Addressed in 472d6593. Thanks for the insight.

File: zend_abstract_interface/functions/functions.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/functions.h:18]: I've addressed the UBSan errors in e8392df. (Marking this unresolved until hearing thoughts on moving args to the caller from @morrisonlevi.)

File: zend_abstract_interface/functions/tests/CMakeLists.txt (hunk 5 above)
Comment (SammyK) [reply zend_abstract_interface/functions/tests/CMakeLists.txt:9]: As far as I can tell from the [`add_dependencies` docs](https://cmake.org/cmake/help/latest/command/add_dependencies.html) (I couldn't find `add_dependency`), the API only supports targets to resolve build-order dependencies. I converted the stubs dir to a target with `add_custom_target`, but this did not copy the files:

```
add_custom_target(function_stubs SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/stubs)
add_dependencies(functions function_stubs)
```

I was able to copy the files using [`add_custom_command`](https://cmake.org/cmake/help/latest/command/add_custom_command.html):

```
add_custom_command(TARGET functions
                   POST_BUILD
                   COMMAND ${CMAKE_COMMAND} -E copy_directory
                           ${CMAKE_CURRENT_SOURCE_DIR}/stubs
                           ${CMAKE_CURRENT_BINARY_DIR}/stubs)
```

However, this still will not copy the stub files when they are changed. It will only recopy if you make a change to a source file from the target. Effectively it's the same behavior as `COPY`.

There might be a better solution out there, but I haven't been able to find it.

Comment (morrisonlevi) [reply zend_abstract_interface/functions/tests/CMakeLists.txt:9]: We'll just leave this alone for now, as the proper way eludes us.

File: zend_abstract_interface/functions/functions.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (morrisonlevi) [reply zend_abstract_interface/functions/functions.h:18]: If it weren't for UX and minimizing the space for users to get things wrong, I would prefer caller allocated memory over variable length arrays as compiler vendors seem to really dislike it. Notably, MSVC doesn't support VLAs (or if it does it was only recently changed). I think having an upper-bound on argument length as previously suggested would be a decent compromise, as I think we should not intentionally add any more friction to eventually supporting Windows.

[the remark this answers is not in the document]
Comment (SammyK) [reply zend_abstract_interface/functions/functions.h:18]: Thanks for sharing the info. I've updated the args to use a small fixed size array in 5d3566eb.

File: zend_abstract_interface/functions/functions.h [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ *   zval_ptr_dtor(&retval);
+ *
+ * Functions cannot be called outside of a request context so this MUST be
+ * called from within a request context (after RINIT and before RSHUTDOWN).
+ */
+bool zai_call_function(const char *name, size_t name_len, zval *retval, int argc, ...);
```
Comment (bwoebi) [inline zend_abstract_interface/functions/functions.h:25]: do we want to add like ~6 lines of macro hell to avoid manually passing the arg count? Looks to me like such an easy pitfall to mess up. The resulting macro would look similarly to https://github.com/php/php-src/blob/d60bc0e2d91d299c6cb03cb48672479bcffd03a7/Zend/zend_ast.h#L227.

Comment (SammyK) [reply zend_abstract_interface/functions/functions.h:25]: I like this idea. I'll tackle this on Monday. Thanks for the suggestion!

Comment (SammyK) [reply zend_abstract_interface/functions/functions.h]: Thanks for helping me figure out the compiler error (missing `retval` 🤦). This was addressed in 0648f8d and f4e4cdc.

**Review by morrisonlevi — APPROVED** [review]
